### PR TITLE
A first simple Sim3 implementation

### PR DESCRIPTION
**Pull request by tcies**
@rdube ptal (assigned you since you have the most minkindr contribs for an active ASL member)

### CODE REVIEW

File: minkindr/include/kindr/minimal/implementation/quat-sim-transform-inl.h [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+  return QuatSimTransformTemplate<Scalar>(
+      QuatTransformationTemplate<Scalar>(
+          T_A_B_.getRotation().inverse(), -T_A_B_.getRotation().inverseRotate(
+              T_A_B_.getPosition() / scale_A_B_)), 1. / scale_A_B_);
+}
+
+template<typename Scalar>
+std::ostream & operator<<(std::ostream & out,
+                          const QuatSimTransformTemplate<Scalar>& sim_3) {
+  out << "Transform:" << std::endl << sim_3.T_A_B_.getTransformationMatrix() <<
```
Comment (schneith) [inline minkindr/include/kindr/minimal/implementation/quat-sim-transform-inl.h]: Why not include the scale into the transformation matrix? [R,t:0,s]

File: minkindr/include/kindr/minimal/quat-sim-transform.h [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+public:
+  EIGEN_MAKE_ALIGNED_OPERATOR_NEW
+
+  typedef QuatTransformationTemplate<Scalar> Transform;
+  typedef Eigen::Matrix<Scalar, 3, Eigen::Dynamic> Vectors;
+
+  // Creates identity similarity transform.
+  QuatSimTransformTemplate();
+
+  QuatSimTransformTemplate(const Transform& T_A_B, const Scalar scale_A_B_);
```
Comment (schneith) [inline minkindr/include/kindr/minimal/quat-sim-transform.h]: scale_A_B

File: minkindr/include/kindr/minimal/quat-sim-transform.h [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+
+  Vectors operator*(const Vectors& rhs) const;
+
+  QuatSimTransformTemplate<Scalar> inverse() const;
+
+  Transform getTransform() const { return T_A_B_; }
+  Scalar getScale() const { return scale_A_B_; }
+
```
Comment (schneith) [inline minkindr/include/kindr/minimal/quat-sim-transform.h]: a method to return the transformation matrix would be nice

File: minkindr/test/test-sim3.cc [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+  Eigen::Matrix3Xd expected_output(3, 1);
+  expected_output << 2., 1., 2.;
+
+  EXPECT_TRUE(EIGEN_MATRIX_EQUAL_DOUBLE(sim3 * input, expected_output));
+}
+
+TEST(QuatSimTransformTest, Inverse) {
+  const RotationQuaternion q(Eigen::Quaterniond(
+      0.64491714, 0.26382416,  0.51605132,  0.49816637));
+  const Eigen::Vector3d t( 4.67833851,  8.52053031,  6.71796159 );
```
Comment (schneith) [inline minkindr/test/test-sim3.cc]: rm space

File: minkindr/test/test-sim3.cc [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+  const QuatSimTransform sim3(QuatTransformation(q, t), M_PI);
+
+  const Eigen::Matrix3Xd input = Eigen::Matrix3d::Random();
```
Comment (schneith) [inline minkindr/test/test-sim3.cc]: Matrix3d

File: minkindr/include/kindr/minimal/implementation/quat-sim-transform-inl.h [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+namespace kindr {
+namespace minimal {
+
+template <typename Scalar>
+QuatSimTransformTemplate<Scalar>::QuatSimTransformTemplate() : scale_A_B_(1.) {}
+
+template <typename Scalar>
+QuatSimTransformTemplate<Scalar>::QuatSimTransformTemplate(
+    const Transform& T_A_B, const Scalar scale_A_B_)
+    : T_A_B_(T_A_B), scale_A_B_(scale_A_B_) {}
```
Comment (schneith) [inline minkindr/include/kindr/minimal/implementation/quat-sim-transform-inl.h]: check bounds on scale_A_B_

File: minkindr/include/kindr/minimal/implementation/quat-sim-transform-inl.h [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+  return T_A_B_.transformVectorized(scale_A_B_ * rhs);
+}
+
+template <typename Scalar>
+QuatSimTransformTemplate<Scalar>
+QuatSimTransformTemplate<Scalar>::inverse() const {
+  return QuatSimTransformTemplate<Scalar>(
+      QuatTransformationTemplate<Scalar>(
+          T_A_B_.getRotation().inverse(), -T_A_B_.getRotation().inverseRotate(
+              T_A_B_.getPosition() / scale_A_B_)), 1. / scale_A_B_);
```
Comment (schneith) [inline minkindr/include/kindr/minimal/implementation/quat-sim-transform-inl.h]: either check scale_A_B_!=0 or ensure that it cant be initialized to 0.

File: minkindr/include/kindr/minimal/implementation/quat-sim-transform-inl.h (hunk 1 above)
Comment (tcies) [inline minkindr/include/kindr/minimal/implementation/quat-sim-transform-inl.h]: technically, it should be [R * s, t; 0, 1], but IMO that would obscure the scale (right?). [R,t:0,s] would be wrong - [R,t:0, 1/s] would be correct if translation were applied before scaling:
```
x
y       R t       T(xyz)       T(xyz)/s
z   *   0 s   =   s        =   1
1
```
Or did I misunderstand you?

Comment (tcies) [inline minkindr/include/kindr/minimal/implementation/quat-sim-transform-inl.h]: Actually, a quick google reveals that there are two schools of when to apply scale: http://www.ethaneade.org/latex2html/lie/node28.html VS https://github.com/raulmur/ORB_SLAM2/blob/master/src/Sim3Solver.cc#L320 
Hmm, which one is correct?

Comment (tcies) [inline minkindr/include/kindr/minimal/implementation/quat-sim-transform-inl.h]: LSD SLAM also uses [Rs, t; 0, 1], so I'd suggest to stick to that. Also, feels more intuitive to me. @igilitschenski opinions?

File: minkindr/test/test-sim3.cc (hunk 5 above)
Comment (tcies) [inline minkindr/test/test-sim3.cc]: 3Xd intended - that's the input type for operator *. Will add a comment.

File: minkindr/include/kindr/minimal/implementation/quat-sim-transform-inl.h (hunk 6 above)
Comment (tcies) [inline minkindr/include/kindr/minimal/implementation/quat-sim-transform-inl.h]: what do you mean by that?

Comment (schneith) [inline minkindr/include/kindr/minimal/implementation/quat-sim-transform-inl.h]: it should not be 0. otherwise the inverse is unhappy

Comment (tcies) [inline minkindr/include/kindr/minimal/implementation/quat-sim-transform-inl.h]: Ha, didn't even think of < 0, good point!

**Comment by tcies** [comment]
olgty?

**Comment by schneith** [comment]
could you add a short comment about the used scale convention?
then it lgtm

**Comment by tcies** [comment]
ok, thanks for the review!

File: minkindr/include/kindr/minimal/implementation/quat-sim-transform-inl.h (hunk 6 above)
Comment (helenol) [inline minkindr/include/kindr/minimal/implementation/quat-sim-transform-inl.h]: Isn't this variable shadowing? `Scalar scale_A_B`, `scale_A_B_(scale_A_B)` instead?

Comment (tcies) [inline minkindr/include/kindr/minimal/implementation/quat-sim-transform-inl.h]: So e.g. eclipse interestingly makes the right interpretation - maybe an initializer list thing. But I'll definitely change the name :) 
screenshot from 2016-05-10 09 36 54

**Pull request merged**